server/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -379,7 +383,9 @@ private InetSocketAddress bindToPort(final String name, final InetAddress hostAd
         PortsRange portsRange = new PortsRange(port);
         final AtomicReference<Exception> lastException = new AtomicReference<>();
         final AtomicReference<InetSocketAddress> boundSocket = new AtomicReference<>();
-        closeLock.writeLock().lock();
+        if (closeLock.writeLock().tryLock() == false) {
+            throw new IllegalStateException("failed to acquire close-write-lock");
```

Comment:
I think we can't get here without a bug: we can't be concurrently closing since we're synchronized on `AbstractLifecycle#lifecycle` in both places, and we really shouldn't be opening any connections before we've finished starting.

```suggestion
            assert false; // can't be concurrently stopping and mustn't be opening any connections yet
            throw new IllegalStateException("failed to acquire close-write-lock");
```